### PR TITLE
[IRGen] Replace duplicate code with new method

### PR DESCRIPTION
**Pull request by dtweston**
This pull request removes duplicated code from `GlobalPropertyOpt::isVisibleExternally()`, `FunctionLivenessComputation::isVisibleExternally()` and the static function `isAssignableExternally()` in `LetPropertiesOpts.cpp`.

Resolves [SR-6264](https://bugs.swift.org/browse/SR-6264).

### CODE REVIEW

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -2073,6 +2074,25 @@ static AccessLevel getTestableAccess(const ValueDecl *decl) {
   return AccessLevel::Public;
 }
 
+bool ValueDecl::isVisibleExternally(bool isWholeModule) const {
+  AccessLevel access = getEffectiveAccess();
+  SILLinkage linkage;
+  switch (access) {
+    case AccessLevel::Private:
```
Comment (xwu) [inline lib/AST/Decl.cpp]: Formatting nit: ‘case’ should have the same level of indenting as ‘switch’ per standard LLVM/Swift project style.

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -46,6 +46,7 @@
 #include "swift/Basic/Range.h"
 #include "swift/Basic/StringExtras.h"
 #include "swift/Basic/Statistic.h"
+#include "swift/SIL/SILLinkage.h"
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: AST should not depend on SIL. Instead, make this a global function in lib/SIL/SIL.cpp.

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -2073,6 +2074,25 @@ static AccessLevel getTestableAccess(const ValueDecl *decl) {
   return AccessLevel::Public;
 }
 
+bool ValueDecl::isVisibleExternally(bool isWholeModule) const {
+  AccessLevel access = getEffectiveAccess();
+  SILLinkage linkage;
+  switch (access) {
+  case AccessLevel::Private:
+  case AccessLevel::FilePrivate:
+    linkage = SILLinkage::Private;
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: This can just fold to 'return false'

File: lib/AST/Decl.cpp [hunk 4]
```diff
@@ -2073,6 +2074,25 @@ static AccessLevel getTestableAccess(const ValueDecl *decl) {
...
+    break;
+  case AccessLevel::Internal:
+    linkage = SILLinkage::Hidden;
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: This is where you return false if isWholeModule, otherwise true

File: lib/AST/Decl.cpp [hunk 5]
```diff
@@ -2073,6 +2074,25 @@ static AccessLevel getTestableAccess(const ValueDecl *decl) {
...
+    break;
+  case AccessLevel::Public:
+  case AccessLevel::Open:
+    linkage = SILLinkage::Public;
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: This is just true

**Review by slavapestov — CHANGES_REQUESTED** [review]

File: lib/AST/Decl.cpp (hunk 2 above)
Comment (dtweston) [reply lib/AST/Decl.cpp]: Ok, that makes sense. I'll rework this.

Comment (dtweston) [reply lib/AST/Decl.cpp]: I force updated this change and moved the implementation to lib/SIL/SIL.cpp. include/SIL/SILModule.h seemed to be the best place for this declaration to live, but would it make sense for this to be a member function on SILModule directly?

**Comment by dtweston** [comment]
@slavapestov, what do you think of the latest version?

File: lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp [hunk 6]
```diff
@@ -257,18 +257,8 @@ static bool isDefaultCaseKnown(ClassHierarchyAnalysis *CHA,
     return false;
 
   // Only consider 'private' members, unless we are in whole-module compilation.
-  switch (CD->getEffectiveAccess()) {
-  case AccessLevel::Open:
+  if (isDeclVisibleExternally(CD, &AI.getModule()))
     return false;
-  case AccessLevel::Public:
```
Comment (dtweston) [inline lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp]: The comment says we should only consider private members, but it treats public members the same as internal members, which makes me think I'm possible misunderstanding this.

File: lib/SILOptimizer/Utils/Devirtualize.cpp [hunk 7]
```diff
@@ -192,19 +182,6 @@ static bool isKnownFinalClass(ClassDecl *CD, SILModule &M,
...
-      case AccessLevel::Public:
-      case AccessLevel::Internal:
-        if (!M.isWholeModule())
-          return false;
-        break;
-      case AccessLevel::FilePrivate:
-      case AccessLevel::Private:
-        break;
-      }
-
```
Comment (dtweston) [inline lib/SILOptimizer/Utils/Devirtualize.cpp]: The above code also appeared on lines 174 to 185, so it seems redundant to duplicate them here.

File: lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp (hunk 6 above)
Comment (slavapestov) [reply lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp]: It's probably just an oversight.

**Review by slavapestov — APPROVED** [review]

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@dtweston Thank you!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp (hunk 6 above)
Comment (dtweston) [reply lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp]: Ok. I think something in this later change broke a SILOptimizer test (global_property_opt.sil), but I wasn't yet able to figure out what it meant.

Here's a gist with the difference: https://gist.github.com/dtweston/1ce13a14a0cadbfe2088aa647ee7a891

File: lib/SIL/SIL.cpp [hunk 8]
```diff
@@ -115,3 +115,18 @@ swift::getLinkageForProtocolConformance(const NormalProtocolConformance *C,
       return (definition ? SILLinkage::Public : SILLinkage::PublicExternal);
   }
 }
+
+bool isDeclVisibleExternally(const ValueDecl *decl, const SILModule *module) {
```
Comment (slavapestov) [inline lib/SIL/SIL.cpp]: I think you to qualify the definition here, like `bool swift::isDeclVisibleExternally()`. CI failed to compile this PR with this error:

```
Undefined symbols for architecture x86_64:
  "swift::isDeclVisibleExternally(swift::ValueDecl const*, swift::SILModule const*)", referenced from:
      (anonymous namespace)::SILGlobalOpt::run() in libswiftSILOptimizer.a(GlobalOpt.cpp.o)
      (anonymous namespace)::LetPropertiesOptPass::run() in libswiftSILOptimizer.a(LetPropertiesOpts.cpp.o)
      (anonymous namespace)::LetPropertiesOpt::isConstantLetProperty(swift::VarDecl*) in libswiftSILOptimizer.a(LetPropertiesOpts.cpp.o)
      (anonymous namespace)::GlobalPropertyOpt::getFieldEntry(swift::VarDecl*) in libswiftSILOptimizer.a(GlobalPropertyOpt.cpp.o)
      (anonymous namespace)::DeadFunctionElimination::findAnchorsInTables() in libswiftSILOptimizer.a(DeadFunctionElimination.cpp.o)
      isDefaultCaseKnown(swift::ClassHierarchyAnalysis*, swift::FullApplySite, swift::ClassDecl*, llvm::SmallVector<swift::ClassDecl*, 8u>&) in libswiftSILOptimizer.a(SpeculativeDevirtualizer.cpp.o)
      isKnownFinalClass(swift::ClassDecl*, swift::SILModule&, swift::ClassHierarchyAnalysis*) in libswiftSILOptimizer.a(Devirtualize.cpp.o)
      ...
ld: symbol(s) not found for architecture x86_64
```

Comment (dtweston) [reply lib/SIL/SIL.cpp]: I'm looking into this.

Comment (dtweston) [reply lib/SIL/SIL.cpp]: Fixed and rebased.

**Comment by dtweston** [comment]
@swift-ci Please smoke test

File: lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp (hunk 6 above)
Comment (dtweston) [reply lib/SILOptimizer/Transforms/SpeculativeDevirtualizer.cpp]: This test breakage has gone away, so I think everything's fine now.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
Hi @dtweston, it looks like the devirt-speculate test failure was real -- do you mind investigating?

**Comment by jrose-apple** [comment]
Closing due to age, but someone else can pick it up.

**Pull request closed without merging**